jujubundlelib/validate.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import unicode_literals
+
+import collections
+import re
+
+import models
+import utils
+
+
+class BundleValidator(object):
+
+    def __init__(self, bundle):
+        self.bundle = bundle
+        self._errors = []
+
+    def add_error(self, error):
+        self._errors.append(error)
+
+    def errors(self):
+        return self._errors
+
+    def _is_legacy_bundle(self):
+        return utils.is_legacy_bundle(self.bundle)
+
+
+def validate(bundle):
+    validator = BundleValidator(bundle)
+    machine_ref_counts = {}
+    for machine in validator.bundle.get('machines', {}).keys():
```

Comment:
We need to first validate that bundle is a dict-like object, otherwise bundle.get would return an AttributeError.